spring-context-support/src/main/java/org/springframework/cache/transaction/TransactionAwareCacheDecorator.java

```diff
@@ -79,18 +111,50 @@ public Object getNativeCache() {
 	@Override
 	@Nullable
 	public ValueWrapper get(Object key) {
-		return this.targetCache.get(key);
+		final var transactionState = getTransactionState();
+		final var current = transactionState == null ? null : transactionState.get(key);
+		if (current != null) {
+			return convert(current);
+		} else {
+			return this.targetCache.get(key);
+		}
 	}
 
 	@Override
 	@Nullable
 	public <T> T get(Object key, @Nullable Class<T> type) {
-		return this.targetCache.get(key, type);
+		final var transactionState = getTransactionState();
+		final var wrapper = transactionState == null ? null : transactionState.get(key);
+		if (wrapper != null) {
+			// Unwrap
+			final var value = convertValue(wrapper);
+			return cast(value, type);
```

Comment:
Potential NPE, consider handling the case where convertValue(wrapper) returns null before calling cast(value, type).